File: vod-toolkits/playauth-generate/Java/src/main/java/com/aliyun/vod/playauth/VodPlayAuthToken.java
package com.aliyun.vod.playauth;

import com.alibaba.fastjson.JSONObject;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.exceptions.JWTCreationException;
import com.auth0.jwt.exceptions.JWTVerificationException;
import org.apache.commons.lang3.StringUtils;

import java.util.HashMap;
import java.util.Map;

/**
 * 阿里云视频点播播放授权 Token 工具类
 * <p>
 * 提供 JWT 格式的播放授权 Token 生成与验证功能
 * <ul>
 *   <li>生成播放授权 Token</li>
 *   <li>验证 Token 有效性与签名</li>
 *   <li>解析 Token 的 Header 和 Payload</li>
 * </ul>
 * <p>
 * <b>重要提示：</b>使用 {@code vid + JWTPlayAuth}（字段名：{@code playAuth}）进行播放时，
 * 客户端播放器 SDK 版本需要满足 {@code >= 7.10.0}，否则无法正常完成播放鉴权。
 * </p>
 *
 * @author fulin
 * @date 2025/9/12 14:28
 */
public final class VodPlayAuthToken {

    /**
     * 私有构造函数，防止实例化
     */
    private VodPlayAuthToken() {
    }

    /**
     * 生成播放授权 Token（使用默认配置）
     * <p>
     * 使用 {@link VodJwtConstants} 中定义的默认参数生成 Token
     *
     * @param videoId 视频 ID，不能为空
     * @return JWT Token 字符串，生成失败返回 {@code null}
     */
    public static String generateToken(String videoId) {
        return generateToken(
                videoId,
                VodJwtConstants.DEFAULT_APP_ID,
                VodJwtConstants.DEFAULT_PLAY_KEY,
                VodJwtConstants.DEFAULT_REGION_ID,
                VodJwtConstants.EXPIRED_TIME_MILLS
        );
    }

    /**
     * 生成播放授权 Token（自定义配置）
     * <p>
     * Token Payload 包含以下字段：
     * <ul>
     *   <li>appId - 应用标识</li>
     *   <li>videoId - 视频 ID</li>
     *   <li>currentTimeStamp - 签发时间戳（毫秒）</li>
     *   <li>expireTimeStamp - 过期时间戳（毫秒）</li>
     *   <li>regionId - 地域标识</li>
     *   <li>playContentInfo - 播放配置信息</li>
     * </ul>
     *
     * @param videoId          视频 ID，不能为空
     * @param appId            应用 ID，不能为空
     * @param playKey          播放密钥，不能为空
     * @param regionId         地域标识，不能为空
     * @param expiredTimeMills 过期时间（毫秒），必须大于 0
     * @return JWT Token 字符串，生成失败返回 {@code null}
     */
    public static String generateToken(String videoId, String appId, String playKey,
                                       String regionId, long expiredTimeMills) {
        long currentTimeStamp = System.currentTimeMillis();
        long expireTimeStamp = currentTimeStamp + expiredTimeMills;

        // 播放内容配置信息
        // 注意：PlayContentInfo 接口入参与原接口 GetPlayInfo 一致，@清邻
        Map<String, Object> playContentInfo = new HashMap<>();
        // 视频格式：MP4
        playContentInfo.put("formats", "mp4");
        // 播放地址的过期时间，默认1小时，可自定义为 30分钟（1800秒）
        playContentInfo.put("authTimeout", 1800L);
        // 流类型：视频流
        playContentInfo.put("streamType", "video");

        try {
            Algorithm algorithm = Algorithm.HMAC256(playKey);
            return JWT.create()
                    .withClaim("appId", appId)
                    .withClaim("videoId", videoId)
                    .withClaim("currentTimeStamp", currentTimeStamp)
                    .withClaim("expireTimeStamp", expireTimeStamp)
                    .withClaim("regionId", regionId)
                    .withClaim("playContentInfo", playContentInfo)
                    .sign(algorithm);
        } catch (JWTCreationException e) {
            System.err.println("JWT Token 创建失败: " + e.getMessage());
            return null;
        }
    }

    /**
     * 验证播放授权 Token（使用默认密钥）
     * <p>
     * 验证流程：
     * <ol>
     *   <li>Token 格式校验（三段式结构）</li>
     *   <li>签名验证</li>
     *   <li>时间戳校验（防重放攻击、过期检查）</li>
     * </ol>
     *
     * @param token 待验证的 JWT Token
     * @return {@code true} 验证通过，{@code false} 验证失败
     */
    public static boolean verifyToken(String token) {
        return verifyToken(token, VodJwtConstants.DEFAULT_PLAY_KEY);
    }

    /**
     * 验证播放授权 Token（自定义密钥）
     *
     * @param token   待验证的 JWT Token
     * @param playKey 播放密钥
     * @return {@code true} 验证通过，{@code false} 验证失败
     */
    public static boolean verifyToken(String token, String playKey) {
        try {
            // 1. Token 格式校验
            String[] parts = StringUtils.split(token, ".");
            if (parts == null || parts.length != 3) {
                System.out.println("Token 格式无效：应为 Header.Payload.Signature 格式");
                return false;
            }

            // 2. 解析 Payload
            String payload = Base64Utils.decodeBase64(parts[1])
                    .replace("\n", "")
                    .replace("\r", "");
            JSONObject jsonObject = JSONObject.parseObject(payload);

            // 3. 签名验证
            Algorithm algorithm = Algorithm.HMAC256(playKey);
            JWT.require(algorithm).build().verify(token);

            // 4. 时间戳校验
            Long currentTimeStamp = jsonObject.getLong("currentTimeStamp");
            if (currentTimeStamp == null || currentTimeStamp > System.currentTimeMillis()) {
                System.out.println("签发时间无效：Token 签发时间不能晚于当前时间");
                return false;
            }

            Long expireTimeStamp = jsonObject.getLong("expireTimeStamp");
            if (expireTimeStamp != null && expireTimeStamp < System.currentTimeMillis()) {
                System.out.println("Token 已过期：当前时间超过了 Token 的有效期");
                return false;
            }

            return true;

        } catch (JWTVerificationException e) {
            System.err.println("JWT 验证失败: " + e.getMessage());
        } catch (Exception e) {
            System.err.println("Token 处理异常: " + e.getMessage());
        }

        return false;
    }

    /**
     * 解析 Token 的 Payload 部分
     *
     * @param token JWT Token
     * @return Payload JSON 字符串，解析失败返回 {@code null}
     */
    public static String parsePayload(String token) {
        try {
            String[] parts = StringUtils.split(token, ".");
            if (parts == null || parts.length != 3) {
                return null;
            }
            return Base64Utils.decodeBase64(parts[1])
                    .replace("\n", "")
                    .replace("\r", "");
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * 解析 Token 的 Header 部分
     *
     * @param token JWT Token
     * @return Header JSON 字符串，解析失败返回 {@code null}
     */
    public static String parseHeader(String token) {
        try {
            String[] parts = StringUtils.split(token, ".");
            if (parts == null || parts.length != 3) {
                return null;
            }
            return Base64Utils.decodeBase64(parts[0])
                    .replace("\n", "")
                    .replace("\r", "");
        } catch (Exception e) {
            return null;
        }
    }
}

File: vod-toolkits/playauth-generate/Go/playauth/vod_play_auth_token.go
package playauth

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// PlayContentInfo 播放内容配置信息
type PlayContentInfo struct {
	Formats     string `json:"formats"`
	AuthTimeout int64  `json:"authTimeout"` // 单位：秒
	StreamType  string `json:"streamType"`
}

// CustomClaims 自定义 JWT Claims
// 嵌入 jwt.RegisteredClaims 以满足 jwt.Claims 接口
type CustomClaims struct {
	AppID            string          `json:"appId"`
	VideoID          string          `json:"videoId"`
	CurrentTimeStamp int64           `json:"currentTimeStamp"`
	ExpireTimeStamp  int64           `json:"expireTimeStamp"`
	RegionID         string          `json:"regionId"`
	PlayContentInfo  PlayContentInfo `json:"playContentInfo"`
	jwt.RegisteredClaims
}

// 阿里云视频点播播放授权 Token 工具类
// 提供 JWT 格式的播放授权 Token 生成与验证功能:
//   - 生成播放授权 Token
//   - 验证 Token 有效性与签名
//   - 解析 Token 的 Header 和 Payload
//
// 重要提示: 使用 vid + JWTPlayAuth（字段名: playAuth）进行播放时，
// 客户端播放器 SDK 版本需要满足 >= 7.10.0，否则无法正常完成播放鉴权。
//
// @author wyq
// @date 2025/9/12 14:28

// GenerateToken 生成播放授权 Token（使用默认配置）
// 使用 vod_jwt_constants 中定义的默认参数生成 Token
func GenerateToken(videoID string) (string, error) {
	return GenerateTokenWithConfig(
		videoID,
		DefaultAppID,
		DefaultPlayKey,
		DefaultRegionID,
		ExpiredTimeMills,
	)
}

// GenerateTokenWithConfig 生成播放授权 Token（自定义配置）
// Token Payload 包含以下字段:
//   - appId: 应用标识
//   - videoId: 视频 ID
//   - currentTimeStamp: 签发时间戳（毫秒）
//   - expireTimeStamp: 过期时间戳（毫秒）
//   - regionId: 地域标识
//   - playContentInfo: 播放配置信息
func GenerateTokenWithConfig(videoID, appID, playKey, regionID string, expiredTimeMills int64) (string, error) {
	currentTime := time.Now().UnixMilli()
	expireTime := currentTime + expiredTimeMills

	// 播放内容配置信息
	// 注意: PlayContentInfo 接口入参与原接口 GetPlayInfo 一致
	playContentInfo := PlayContentInfo{
		Formats:     "mp4",
		AuthTimeout: 1800,
		StreamType:  "video",
	}

	claims := CustomClaims{
		AppID:            appID,
		VideoID:          videoID,
		CurrentTimeStamp: currentTime,
		ExpireTimeStamp:  expireTime,
		RegionID:         regionID,
		PlayContentInfo:  playContentInfo,
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	tokenString, err := token.SignedString([]byte(playKey))
	if err != nil {
		return "", err
	}
	return tokenString, nil
}

// VerifyToken 验证播放授权 Token（使用默认密钥）
// 验证流程:
//   - Token 格式校验（三段式结构）
//   - 签名验证
//   - 时间戳校验（防重放攻击、过期检查）
func VerifyToken(tokenStr string) bool {
	return VerifyTokenWithKey(tokenStr, DefaultPlayKey)
}

// VerifyTokenWithKey 验证播放授权 Token（自定义密钥）
func VerifyTokenWithKey(tokenStr, playKey string) bool {
	// 1. Token 格式校验
	parts := strings.Split(tokenStr, ".")
	if len(parts) != 3 {
		fmt.Println("Token 格式无效：应为 Header.Payload.Signature 格式")
		return false
	}

	// 2. 解析 Payload
	payloadB64 := parts[1]
	padding := len(payloadB64) % 4
	if padding > 0 {
		payloadB64 += strings.Repeat("=", 4-padding)
	}
	payloadBytes, err := base64.StdEncoding.DecodeString(payloadB64)
	if err != nil {
		fmt.Printf("Payload Base64 解码失败: %v\n", err)
		return false
	}

	var payloadMap map[string]interface{}
	if err := json.Unmarshal(payloadBytes, &payloadMap); err != nil {
		fmt.Printf("Payload JSON 解析失败: %v\n", err)
		return false
	}

	// 3. 签名验证
	parser := new(jwt.Parser)
	_, err = parser.ParseWithClaims(tokenStr, &CustomClaims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(playKey), nil
	})
	if err != nil {
		fmt.Printf("JWT 验证失败: %v\n", err)
		return false
	}

	// 4. 时间戳校验
	now := time.Now().UnixMilli()

	currentTimeStamp, ok := payloadMap["currentTimeStamp"].(float64)
	if !ok || int64(currentTimeStamp) > now {
		fmt.Println("签发时间无效：Token 签发时间不能晚于当前时间")
		return false
	}

	expireTimeStamp, ok := payloadMap["expireTimeStamp"].(float64)
	if ok && int64(expireTimeStamp) < now {
		fmt.Println("Token 已过期：当前时间超过了 Token 的有效期")
		return false
	}

	return true
}

// ParsePayload 解析 Token 的 Payload 部分
func ParsePayload(tokenStr string) string {
	parts := strings.Split(tokenStr, ".")
	if len(parts) != 3 {
		return ""
	}
	payloadB64 := parts[1]
	padding := len(payloadB64) % 4
	if padding > 0 {
		payloadB64 += strings.Repeat("=", 4-padding)
	}
	payloadBytes, err := base64.StdEncoding.DecodeString(payloadB64)
	if err != nil {
		return ""
	}
	return string(payloadBytes)
}

// ParseHeader 解析 Token 的 Header 部分
func ParseHeader(tokenStr string) string {
	parts := strings.Split(tokenStr, ".")
	if len(parts) != 3 {
		return ""
	}
	headerB64 := parts[0]
	padding := len(headerB64) % 4
	if padding > 0 {
		headerB64 += strings.Repeat("=", 4-padding)
	}
	headerBytes, err := base64.StdEncoding.DecodeString(headerB64)
	if err != nil {
		return ""
	}
	return string(headerBytes)
}
